# En este notebook empezaremos haceidno un analisis exploratorio y un datacleaning

Para empezar importemos las dependencias, definamos la ruta al dataset y luego, lo convertremos en un dataframe.

In [1]:
# instalamos el modulo pandas
!pip install pandas


[notice] A new release of pip is available: 23.2.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
# importamos dependencias
import pandas as pd
from pathlib import Path

In [3]:
# defininmos el path
path = Path(r'C:\Users\antim\OneDrive\Escritorio\poryectos\proyecto_videojuegos\data\vgsales.csv')
path

WindowsPath('C:/Users/antim/OneDrive/Escritorio/poryectos/proyecto_videojuegos/data/vgsales.csv')

In [4]:
# Ahora lo convrtimos en un dataframe
data = pd.read_csv(path)

# Veamos si se cargo bien
data.head(4), data.shape

(   Rank               Name Platform    Year     Genre Publisher  NA_Sales  \
 0     1         Wii Sports      Wii  2006.0    Sports  Nintendo     41.49   
 1     2  Super Mario Bros.      NES  1985.0  Platform  Nintendo     29.08   
 2     3     Mario Kart Wii      Wii  2008.0    Racing  Nintendo     15.85   
 3     4  Wii Sports Resort      Wii  2009.0    Sports  Nintendo     15.75   
 
    EU_Sales  JP_Sales  Other_Sales  Global_Sales  
 0     29.02      3.77         8.46         82.74  
 1      3.58      6.81         0.77         40.24  
 2     12.88      3.79         3.31         35.82  
 3     11.01      3.28         2.96         33.00  ,
 (16598, 11))

# EDA

El analisis exploratrio (EDA), se puede dividir en dos etapas:

### EDA Diagnostico:
    Lo realizaremos antes de limpiar el dataset, y la idea es explorar los datos.

### EDA Estrategico:
    Lo realizaremos luego de limpiar el dataset.
 
# Pasos de EDA diagnostico.
    Uilizaremos metodos como `info()`, `describe()`, etc.

Nos plantearmos las siguientes preguntas para el EDA daignostico:
    - Concentracion: Comprar el maximo de recaudacion con el percentil del 75%
    - Cuantos generos existen en el dataset 
    - Sesgo temporal: Miremos las fechas donde empieza y donde termina.    

In [5]:
# Este metodo nos permite ver un resumen del dataframe incluido tipos de datos
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16598 non-null  int64  
 1   Name          16598 non-null  str    
 2   Platform      16598 non-null  str    
 3   Year          16327 non-null  float64
 4   Genre         16598 non-null  str    
 5   Publisher     16540 non-null  str    
 6   NA_Sales      16598 non-null  float64
 7   EU_Sales      16598 non-null  float64
 8   JP_Sales      16598 non-null  float64
 9   Other_Sales   16598 non-null  float64
 10  Global_Sales  16598 non-null  float64
dtypes: float64(6), int64(1), str(4)
memory usage: 2.1 MB


podemos que tenemos datos de tipo int, str y float.

Ahora utilizemos el metodo `describe()` -> Recordemos que este metodo solo sirve para columnas numericas.

In [6]:
data.describe()

,Rank,Year,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales
count,16598.000000,16327.000000,16598.000000,16598.000000,16598.000000,16598.000000,16598.000000
mean,8300.605254,2006.406443,0.264667,0.146652,0.077782,0.048063,0.537441
std,4791.853933,5.828981,0.816683,0.505351,0.309291,0.188588,1.555028
min,1.000000,1980.000000,0.000000,0.000000,0.000000,0.000000,0.010000
25%,4151.250000,2003.000000,0.000000,0.000000,0.000000,0.000000,0.060000
50%,8300.500000,2007.000000,0.080000,0.020000,0.000000,0.010000,0.170000
75%,12449.750000,2010.000000,0.240000,0.110000,0.040000,0.040000,0.470000
max,16600.000000,2020.000000,41.490000,29.020000,10.220000,10.570000,82.740000


Con esto podemos sacara muchas conclusiones, como que el year de comienzo es 1980, y el ultimo year es 2020.

Tambienn es importante no scar conclusiones apresuradas sobre los datos, ya que estops aun no estan limpios.

Ahora que tipos de plataformas hay y luego de generos.

In [7]:
# generos totales
generos = data["Genre"].unique()
# plataformas totales
plataformas = data["Platform"].unique()

print(f"Las plataformas son ({plataformas})\nY los generos son ({generos})")

Las plataformas son (<ArrowStringArray>
[ 'Wii',  'NES',   'GB',   'DS', 'X360',  'PS3',  'PS2', 'SNES',  'GBA',
  '3DS',  'PS4',  'N64',   'PS',   'XB',   'PC', '2600',  'PSP', 'XOne',
   'GC', 'WiiU',  'GEN',   'DC',  'PSV',  'SAT',  'SCD',   'WS',   'NG',
 'TG16',  '3DO',   'GG', 'PCFX']
Length: 31, dtype: str)
Y los generos son (<ArrowStringArray>
[      'Sports',     'Platform',       'Racing', 'Role-Playing',
       'Puzzle',         'Misc',      'Shooter',   'Simulation',
       'Action',     'Fighting',    'Adventure',     'Strategy']
Length: 12, dtype: str)


Ahora sabemos que consolas y generos hay ademas de su cantidad.

Veamos como se ven las primeras y ultimas filas.

In [8]:
data.head(5)

,Rank,Name,Platform,Year,Genre,Publisher,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales
0,1,Wii Sports,Wii,2006.0,Sports,Nintendo,41.49,29.02,3.77,8.46,82.74
1,2,Super Mario Bros.,NES,1985.0,Platform,Nintendo,29.08,3.58,6.81,0.77,40.24
2,3,Mario Kart Wii,Wii,2008.0,Racing,Nintendo,15.85,12.88,3.79,3.31,35.82
3,4,Wii Sports Resort,Wii,2009.0,Sports,Nintendo,15.75,11.01,3.28,2.96,33.00
4,5,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,Nintendo,11.27,8.89,10.22,1.00,31.37


In [9]:
data.tail(5)

,Rank,Name,Platform,Year,Genre,Publisher,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales
16593,16596,Woody Woodpecker in Crazy Castle 5,GBA,2002.0,Platform,Kemco,0.01,0.00,0.0,0.0,0.01
16594,16597,Men in Black II: Alien Escape,GC,2003.0,Shooter,Infogrames,0.01,0.00,0.0,0.0,0.01
16595,16598,SCORE International Baja 1000: The Official Game,PS2,2008.0,Racing,Activision,0.00,0.00,0.0,0.0,0.01
16596,16599,Know How 2,DS,2010.0,Puzzle,7G//AMES,0.00,0.01,0.0,0.0,0.01
16597,16600,Spirits & Spells,GBA,2003.0,Platform,Wanadoo,0.01,0.00,0.0,0.0,0.01


Habiendo hecho este primer analisis, podeos responder:

- El valor del percentil 75% de la columna `global_sales` es de `0.470000` .

- Hay una amplia variedad de generos siendo que hay un total de 12.

- El rango temporal del dataset va desde `1980` a `2020` .

# Limpieza de datos

Ahora limpiaremos los datos y luego realizaremos el EDA Estrategico.

Limpieza : 

0. Columnas a lowerCase:
    Antes de comenzar con la limpieza de datos en si, primero limpiaremos las columnas haceindo que estas sean minuscula en vez de mayuscula.

1. Gestion de datos nulos en la columna `'year'`
    Lo que haremos sera eliminar los datos nulos de la columna 'year', ya que la cantidad de datos nulos solo reepresenta el 2%, y en estos casos es mejor que los datos tengan menos ruido y no cambiarlos por valores como la media o mediana. Para esto utilizaremos el metodo `.dropna()`.\

2. Intercambio de datos nulo en la columna `'publisher'`
    Cambiaremos los datos nulos de la coumna 'publisher' con el string 'Unknown', para no perder las filas que no tienen publisher. Esto lo haremos con el metodo `.fillna()`

3. Duplicados
    Eliminaremos aquellos registros duplicados donde el name y la plataform sea la misma. Para esto utilizaremos el metodo `.drop_duplicates(subset=[])`

4. Casting de tipos de dato
    Una vez eliminemos lo registros con year nulo, podemos hacer un casteo de la columna year podemos pasar la columna de tipo float a int.

In [10]:
# cambiamos el atributo publico columns a lower(), !!! importante utilizar el .str para wue s3ea un string
data.columns = data.columns.str.lower()

data.columns

Index(['rank', 'name', 'platform', 'year', 'genre', 'publisher', 'na_sales',
       'eu_sales', 'jp_sales', 'other_sales', 'global_sales'],
      dtype='str')

ya tenemos las columnas en minuscula sigamos con la getsion de nulos con la columna 'year'.

Utilizaremos el metodo `dropna()` --> recordar que este metodo es un metodo inmutable por lo que no modifica el dataframe original.

In [11]:
# primero veamos la cantidad de valores nulos que tenemos en la columna year
print(data['year'].isnull().sum())

# eliminamos los valores nulos de la columna
data = data.dropna(subset="year")

print(data['year'].isnull().sum())

271
0


Vemos como pasamos de 271 valores nulos a 0. 

Ahora lo que haremos sera rempplzar los datos nulos de la columna 'publisher' por el str 'Unknown'

In [12]:
# primero veamos los valores nulos de la columna 
print(data['publisher'].isnull().sum())

# cambiamos los valores nulos por 'Unknown'
data['publisher'] = data['publisher'].fillna('Unknown') # es importaante seleccionar que columna queremos remplzar

# veamos que valores nulos quedaron
print(data['publisher'].isnull().sum())

36
0


Ahora vemos como no tenemos mas valores nulos dentro de la columna 'publisher', sino que ahora aparecera el dato 'Unkonwn'

Ahora eliminaremos los registros dupliocados , estos los identiicaremos si su columna 'name' y 'plataaform' son iguales, nos quedaremos con el primer registro encontrado, eliminando todos los demas.

In [13]:
# primero contemos cuantos registros tienen al menos una concidencia
print(f"Cantidad de duplicados: {data[['name', 'platform']].duplicated().sum()}")
# contemos cuantos duplicados tienen estos registros
print(f"\nJuegos y plataformas duplicados:\n{data[['name', 'platform']].value_counts()[data[['name', 'platform']].value_counts() > 1]}\n")

# eliminamos los duplicados utlizando el drop_duplicated -> es un metodo inmutable, por lo que hay que hacer una asiganacion
data = data.drop_duplicates(subset=['name', 'platform'], keep='first') # utilizamos el keep para definir con que registro nos queremos quedar

# imprimammos cuantos quedaron
print(f"Cantidad de duplicados: {data[['name', 'platform']].duplicated().sum()}")

Cantidad de duplicados: 3

Juegos y plataformas duplicados:
name                         platform
Madden NFL 13                PS3         2
Need for Speed: Most Wanted  X360        2
                             PC          2
Name: count, dtype: int64

Cantidad de duplicados: 0


Ahora cambiaremos el tipo de dato de las columnas utilizando el metodo `astype`.

In [14]:
data['year'] = data['year'].astype(int)

data['year'].dtype

dtype('int64')

Por timo veamos si tenemos nulos silenciosos en alguno de nuestros registros en la columna 'name'

In [15]:
# Esto busca filas donde el nombre, tras quitar espacios, quede vacío
nombres_vacios = data[data['name'].str.strip() == ""]
print(f"Nombres que son solo espacios: {len(nombres_vacios)}")

Nombres que son solo espacios: 0


Podemos ver que no tenemos valores de name que sean nulos silenciosos.

Una vez finalisadas estas tareas terminamos con la limpieza, y podemos pasar a la siguiente fase, la cual estara en el notebook numero 2.

Antes de empezar con el notebook 2, pasaremos nuestro dataframe a un archivo tipo parquet el cual podamos exportar. Por que parqeut? --> porque este tipo de archivo puede guardar los dtypes de las columnas, y ademas es mucho mas ligero.

In [16]:
# instalamos una extension para converit el dataframe a un parquet.
!pip install pyarrow


[notice] A new release of pip is available: 23.2.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [17]:
# primero definimos una ruta para guardar el .paqrquet
path = r"C:\Users\antim\OneDrive\Escritorio\poryectos\proyecto_videojuegos\data\dataset_limpio.parquet"

#guardamos el archivo
data.to_parquet(path)

Listo ya tenemos nuestro archivo .parquet en el directorio.